VSC571_Config_Manager_Test_Class      VSC578_Test_Manager_Class

muzman123 opened the pull request:
Created a Test Manager Class and a Config Manager Test Class
Edited CMakeLists.txt to integrate Qt Testing

UniqueNameHere21 commented on CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@ set(CMAKE_AUTOUIC ON)
 configure_file(${CMAKE_SOURCE_DIR}/config.ini.example ${CMAKE_BINARY_DIR}/config.ini COPYONLY)
 
-find_package(Qt6 6.2 REQUIRED COMPONENTS Core Gui Qml Quick Mqtt SerialPort)
+find_package(Qt6 6.2 REQUIRED COMPONENTS Core Gui Qml Test Quick Mqtt SerialPort)
```
Find a way to configure this so Test libs are not required for our prod build.

We do not want to package unit tests onto the pi when we have limited space to work with

UniqueNameHere21 commented on src/Config/ConfigManager.h:
```diff
@@ -50,6 +50,9 @@ class ConfigManager {
     bool loggingEnabled_;
+
+    // To allow test class access to private fields
+    friend class ConfigManagerTest;
```
No reason to give test class access to private fields, should only be testing public methods.

UniqueNameHere21 commented on src/Tests/ConfigManagerTest.cpp:
```diff
@@ -0,0 +1,90 @@
+    // 3. Instantiate ConfigManager with the partial config
+    ConfigManager manager(configPath);
+
+    // 4. Verify that the provided keys override defaults
```
In this case you should also verify the the items not in the partial_config remained the default values

UniqueNameHere21 commented on src/Tests/ConfigManagerTest.cpp:
```diff
@@ -0,0 +1,90 @@
+// Passes if default values are used when no config file exists
+void ConfigManagerTest::testNoConfigFile()
+{
+    QString nonExistentPath = QDir::temp().filePath("this_config_does_not_exist.ini");
```
this_config_does_not_exist.ini -> ThisConfigDoesNotExist.ini

No underscores in file naming for consistency

UniqueNameHere21 commented on src/Tests/ConfigManagerTest.cpp:
```diff
@@ -0,0 +1,90 @@
+    QTemporaryDir tempDir;
+    QVERIFY2(tempDir.isValid(), "Failed to create temporary directory for partial config test.");
+
+    QString configPath = tempDir.filePath("partial_config.ini");
```
file name

UniqueNameHere21 commented on src/Tests/ConfigManagerTest.h:
```diff
@@ -0,0 +1,19 @@
+
+private slots:
+    void testNoConfigFile();
+    void testPartialConfig();
```
Should have 3 test cases:

- No config file found
- Config file with some missing fields (partial config)
- normal config file

UniqueNameHere21 commented on src/main.cpp:
```diff
@@ -7,6 +7,8 @@
 #include <QDir>
 #include <QDebug>
 
+#include "src/Tests/TestManager.h"
```
Test should happen in a separate process from our main application.

An idea for this would to have a class in Tests that we can Run independently from Mercury to run all the unit tests where it can call each Mercury class and test it as required.

UniqueNameHere21 reviewed (changes requested):
Overall looks good and very readable just needs a couple tweaks.

- Configure tests to run independently from Mercury so we don't have to build them into our application
- add missing test case
- Would be very nice if coverage metrics could be gathered as well

Going forward have one branch per jira tracker and following naming convention:
VSCXXX-my-branch-name

muzman123 commented:
- built MercuryTests subproject 
   - added additional test case in ConfigManagerTest
   
- added option to build Mercury without test suite or QtTest modules in parent CMake
- put src code of mercury into Qt Library in CMake (MercuryLib) for cleaner linking to test module

muzman123 commented:
-Each cmake can only be linked to a single main function
      -Therefore each unit test class will now be implemented in its own Qt test subproject for ease of use and accessibility
